### ISSUE TITLE
ListObjects returning IDs not from the chosen type when contextual tuples are involved

### ISSUE THREAD

**Issue by rhamzeh** [openfga/openfga#354]
To reproduce:

```py
type user
type resource
  relations
    define parent as self
    define viewer as viewer from parent 
type folder
  relations
    define allowed as self
    define viewer as allowed
```

With the following tuples:
```yaml
- user: folder:folder-1
  relation: parent
  object: resource:resource-1
```

Note that this issue does not appear when the relation definition is `define viewer as self or allowed` or `define viewer as self`, and it would appear if the relation definition is `define viewer as self and allowed`

To try this out, the sample store: https://play.fga.dev/stores/create/?id=01GJ41921A4W4A3DCN9A5JQEJ8

```sh
curl  -X POST 'https://api.playground.fga.dev/stores/01GJ41921A4W4A3DCN9A5JQEJ8/list-objects' \
-H "Content-Type: application/json" \
-d '{"user": "user:emily", "relation":"viewer","type":"resource","contextual_tuples":{"tuple_keys":[{"user":"user:emily", "relation":"allowed", "object":"folder:folder-1"}]}}' | jq
```

Thanks to Suresh for raising this issue [on discord](https://discord.com/channels/759188666072825867/1042596768031973466)

<img width="641" alt="Screen Shot 2022-11-18 at 5 15 02 PM" src="https://user-images.githubusercontent.com/536147/202812120-ae2a7df3-931d-4c15-bb81-26db4536ae59.png">

**Comment by miparnisari** [comment]
I think the root cause is this

**Comment by jon-whit** [comment]
Fixed in release `v0.3.0`
